### DISCUSSION TITLE
AOT Compatible Questions

### DISCUSSION THREAD

**Discussion by ghost1372** [microsoft/WindowsAppSDK#4719] [Q&A] [answered]
Hi, I am trying to make my libraries compatible with AOT.
In the last call, it was stated that the following tasks should be done for this:

- 1. `<IsAOTCompatible>true</IsAOTCompatible>`
- 2. mark all classes as `partial`
- 3. Replace reflection-based code with statically-typed alternatives:
 Foo.GetType() -> typeof(Foo)
JsonSerializer with SourceGenerationContext
DllImport -> LibraryImport

- 4. Replace assembly-specific code:
Assembly.GetEntryAssembly().Location -> AppContext.BaseDirectory

- 5. Address all AOT/Trim-related warnings

I have done #1, #2, #4 and #5. and I don't have any warnings about AOT/Trim. But I have codes that I have used ` Foo.GetType()` and `DllImport` (I have no warning about them).

Therefore, the question is:
- Are my libraries currently compatible with AOT?
- Do I have to change the mentioned items to the recommended code even though they don't have warnings?
- I tried to change the mentioned items to the recommended code, but it didn't seem possible (for my library), So what happens to that part of the code when the app tries to execute that code?

I will tag the people who may be able to help, I apologize in advance
@Scottj1s 
@codendone 
@Gaoyifei1011

**Comment by Scottj1s** [comment, marked as answer]
@ghost1372 The general answer is that if your code compiles warning free, it should execute error free.  If the type is known statically at compile time (as it would be with FooType.GetType(), but not necessarily with fooInstance.GetType()), then the code is AOT/Trim safe.  As for LibraryImport, the compile may not be warning for DllImport.  @manodasanW, do you remember warnings in gallery that required use LibraryImport?

See also:
https://devblogs.microsoft.com/dotnet/creating-aot-compatible-libraries/

**Comment by Gaoyifei1011** [comment]
> Hi, I am trying to make my libraries compatible with AOT. In the last call, it was stated that the following tasks should be done for this:
> 
> * 1. `<IsAOTCompatible>true</IsAOTCompatible>`
> * 2. mark all classes as `partial`
> * 3. Replace reflection-based code with statically-typed alternatives:
>      Foo.GetType() -> typeof(Foo)
>      JsonSerializer with SourceGenerationContext
>      DllImport -> LibraryImport
> * 4. Replace assembly-specific code:
>      Assembly.GetEntryAssembly().Location -> AppContext.BaseDirectory
> * 5. Address all AOT/Trim-related warnings
> 
> I have done #1, #2, #4 and #5. and I don't have any warnings about AOT/Trim. But I have codes that I have used and (I have no warning about them).` Foo.GetType()``DllImport`
> 
> Therefore, the question is:
> 
> * Are my libraries currently compatible with AOT?
> * Do I have to change the mentioned items to the recommended code even though they don't have warnings?
> * I tried to change the mentioned items to the recommended code, but it didn't seem possible (for my library), So what happens to that part of the code when the app tries to execute that code?
> 
> I will tag the people who may be able to help, I apologize in advance @Scottj1s @codendone @Gaoyifei1011

就我个人而言，如果您在编译过程中没有发现 Visual Studio 给出任何有关裁剪和 AOT 编译时的警告，我个人认为该库就与 AOT 兼容。
当然，如果您仍然对您写的应用程序和第三方库存有疑虑，您可以做如下操作。
1.确保您使用的所有的第三方库都与 AOT 兼容，这个问题不一定能检测到，可能需要您手动测试。（比如目前版本的 Microsoft.Web.WebView2就与 AOT 兼容，编译不存在任何问题，但是不能正常运行使用了 WebView2 库的应用程序）
2.如果您对您写的部分代码存在疑虑，您可以尝试写一个测试函数，调用存在疑虑的代码，尝试在 Debug 模式下开启本机测试来测试相关代码。
（1）在 Debug 模式下开启 AOT 本机编译
```
  <Target Name="AutoPublish" AfterTargets="AfterBuild" Condition="$(Configuration) == 'Debug'">
    <MSBuild Projects="$(MSBuildProjectFullPath)" Targets="Publish" />
  </Target>
```
（2）在调试中需要将调试方法从混合（托管和本机）切换为本机

确保做到上述两点，您可以就做到您的应用程序与 AOT 兼容。


Personally, if you don't see any warnings from Visual Studio about clipping and AOT compilation during the compilation process, I personally think the library is AOT compatible.
Of course, if you still have doubts about the apps and third-party inventory you wrote, you can do the following.
1. Make sure that all the third-party libraries you use are AOT compatible, this issue may not be detected and may require you to test manually. (For example, the current version of Microsoft.Web.WebView2 is compatible with AOT, and there is no problem with compilation, but it does not work properly with applications that use the WebView2 library)
2. If you have doubts about some of the code you wrote, you can try to write a test function, call the code that you have doubts, and try to open native tests in Debug mode to test the relevant code.
(1) Turn on AOT native compilation in Debug mode
```
  <Target Name="AutoPublish" AfterTargets="AfterBuild" Condition="$(Configuration) == 'Debug'">
    <MSBuild Projects="$(MSBuildProjectFullPath)" Targets="Publish" />
  </Target>
```
(2) In debugging, you need to switch the debugging method from hybrid (managed and native) to native

Make sure you do both of the above so that your application is AOT compatible.
